SR-106: New floating-point `description` implementation

tbkka opened the pull request:
This replaces the current implementation of `description` and
`debugDescription` for the standard floating-point types with a new
formatting routine that provides both improved performance and
better results.

Unlike the earlier code based on `sprintf` with a fixed number of
digits, this logic automatically chooses the optimal number of digits,
generating compact output where possible while ensuring round-trip
accuracy. As such, we can now use the exact same output for both `description` and
`debugDescription` (except of course that `debugDescription` provides
full detail for NaNs).

This resolves [SR-106](https://bugs.swift.org/browse/SR-106), [SR-454](https://bugs.swift.org/browse/SR-454), [SR-491](https://bugs.swift.org/browse/SR-491), [SR-3131](https://bugs.swift.org/browse/SR-3131) and other related issues that have complained
about the floating-point `description` and `debugDescription` properties being inexact
and/or inconsistent.

## Ergonomics

With the new code, the REPL generally prints the values you would expect.
```
(swift) 1.1
// r0 : Double = 1.1
(swift) (1.1).description
// r1 : String = "1.1"
(swift) 1e23
// r2 : Double = 1e+23
(swift) print("\(1e23.nextDown) \(1e23) \(1e23.nextUp)")
9.999999999999997e+22 1e+23 1.0000000000000001e+23
(swift) 1.100000000000001
// r3 : Double = 1.100000000000001
(swift) print("\(1.100000000000001)")
1.100000000000001
(swift) 1.0 / 10.0
// r4 : Double = 0.1
(swift) 1.0 / 10.0 + 1.0
// r5 : Double = 1.1
```

In comparison, the previous implementation routinely prints extraneous digits for `debugDescription` (used by the REPL) and omits significant digits in `description` (used by `print`):
```
Welcome to Apple Swift version 4.1 (swiftlang-902.0.38 clang-902.0.31). Type :help for assistance.
  1> 1.1
$R0: Double = 1.1000000000000001
  2> (1.1).description
$R1: String = "1.1"
  3> 1e23
$R2: Double = 9.9999999999999991E+22
  4> print("\(1e23.nextDown) \(1e23) \(1e23.nextUp)")
1e+23 1e+23 1e+23
  5> 1.100000000000001
$R3: Double = 1.100000000000001
  6> print("\(1.100000000000001)")
1.1
  7> 1.0 / 10.0
$R4: Double = 0.10000000000000001
  8> 1.0 / 10.0 + 1.0
$R5: Double = 1.1000000000000001
```

Of course, this only changes how the floating-point numbers are printed.  The actual parsing, storage, and arithmetic operations are unaffected and are still subject to the same rounding issues common to all floating-point arithmetic.

## About the Algorithm

The `SwiftDtoa.c` file here implements a variation of Florian Loitsch' Grisu2
algorithm with changes suggested by Andrysco, Jhala, and Lerner's 2016
paper describing Errol3.

The implementation is:

* Fast.  It uses only fixed-width integer arithmetic and has constant
  memory and time requirements.

* Simple. It is only a little more complex than Loitsch' original
  implementation of Grisu2.  The digit decomposition logic for double is
  less than 300 lines of standard C (half of which is common arithmetic
  support routines).

* Always Accurate. Converting the decimal form back to binary using an
  accurate algorithm (such as Clinger's algorithm) will always yield exactly the
  original binary value.  For the IEEE 754 formats, the round-trip will
  produce exactly the same bit pattern in memory.  This is an essential
  requirement for debugging, logging, and JSON serialization.

* Always Short.  This always selects an accurate result with the minimum
  number of decimal digits.  (So that `1.0 / 10.0` will always print `0.1`.)

* Always Close.  Among all accurate, short results, this always chooses
  the result that is closest to the exact floating-point value. (In case
  of an exact tie, it rounds the last digit even.)

## Performance

The graph below compares the new code (in green) to the performance of three other
popular algorithms.  Note that this graph benchmarks just the underlying C code; the Swift
`description` logic must spend additional effort to allocate a returned String.
* Dragon4 (in yellow) uses variable-length arithmetic in order to provide accurate results regardless of the number of digits requested.  This causes it to become significantly slower when the input has a large positive or negative exponent.  (The other algorithms here do not offer arbitrary numbers of digits, so can use much faster fixed-width arithmetic.)  This is the algorithm commonly used by C standard library implementations of the `printf` family of functions.
* Grisu3 + Dragon4 (in red) uses fast fixed-width arithmetic for 99% of values and falls back to Dragon4 about 1% of the time.   Note the upper whisker that goes to the same height as Dragon4.
* Errol4 (in blue) is a recent algorithm that uses multiple-precision floating-point arithmetic internally.  It is the successor to Errol3 which partly inspired the new Swift implementation.
* The new `SwiftDtoa` implementation (in green) is similar to the Errol algorithms but uses fixed-width integer arithmetic throughout.  This gives it uniform fast performance regardless of the input value.

<img width="619" alt="screen shot 2018-03-23 at 5 53 13 pm" src="https://user-images.githubusercontent.com/21696764/37858541-2b83677c-2ec3-11e8-9c5e-9a7dda50e84f.png">

tbkka committed:
SR-106: New floating-point `description` implementation

This replaces the current implementation of `description` and
`debugDescription` for the standard floating-point types with a new
formatting routine based on a variation of Florian Loitsch' Grisu2
algorithm with changes suggested by Andrysco, Jhala, and Lerner's 2016
paper describing Errol3.

Unlike the earlier code based on `sprintf` with a fixed number of
digits, this version always chooses the optimal number of digits.  As
such, we can now use the exact same output for both `description` and
`debugDescription` (except of course that `debugDescription` provides
full detail for NaNs).

The implementation has been extensively commented; people familiar with
Grisu-style algorithms should find the code easy to understand.

This implementation is:

* Fast.  It uses only fixed-width integer arithmetic and has constant
  memory and time requirements.

* Simple. It is only a little more complex than Loitsch' original
  implementation of Grisu2.  The digit decomposition logic for double is
  less than 300 lines of standard C (half of which is common arithmetic
  support routines).

* Always Accurate. Converting the decimal form back to binary (using an
  accurate algorithm such as Clinger's) will always yield exactly the
  original binary value.  For the IEEE 754 formats, the round-trip will
  produce exactly the same bit pattern in memory.  This is an essential
  requirement for JSON serialization, debugging, and logging.

* Always Short.  This always selects an accurate result with the minimum
  number of decimal digits.  (So that `1.0 / 10.0` will always print
  `0.1`.)

* Always Close.  Among all accurate, short results, this always chooses
  the result that is closest to the exact floating-point value. (In case
  of an exact tie, it rounds the last digit even.)

This resolves SR-106 and related issues that have complained
about the floating-point `description` properties being inexact.

tbkka commented:
@swift-ci Please benchmark

tbkka commented:
I just noticed the bottom label got cut off on the graph.  The horizontal axis is `abs(decimal exponent)` binned in groups of 50.  So the first set of bars includes a sampling of numbers in 1e-50...1e+50, the second set covers 1e-100...1e-50 and 1e50...1e100, etc.  The vertical axis is clock cycles.

tbkka commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -83,15 +83,29 @@ extension ${Self} : CustomStringConvertible {
+    } else if sign == .minus {
+      return "-inf"
+    } else {
+      return "inf"
```
If we short-circuit inf here, we probably don't need to worry about it at the next level down.

tbkka replied:
Removed infinity handling in the next layer down.

tbkka commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -83,15 +83,29 @@ extension ${Self} : CustomStringConvertible {
   /// A textual representation of the value.
   @_inlineable // FIXME(sil-serialize-all)
```
Can anyone explain why this is `@_inlineable`?  This seems to me like a natural ABI boundary, so I'm curious if there was a specific reason for pushing it further down.

milseman replied:
@moiseev  ?

lorentey replied:
Based on the FIXME, Max added this mechanically, to smoothen the transition from previous versions where everything in the stdlib was implicitly inlinable. Attributes marked like this were added in bulk and they can (and should) be removed when we know better.

tbkka replied:
I'll do some perf testing to see if removing this makes a real difference.

tbkka commented on stdlib/public/core/Runtime.swift.gyb:
```diff
@@ -408,37 +408,17 @@ internal func _float${bits}ToString(
-    let significand = value.significandBitPattern
-    if significand == 0 {
-      // Infinity
+  if value.isInfinite {
```
This could be omitted if `inf` is handled at the next layer out.

tbkka replied:
I've removed infinity handling here.  Infinity is now handled in two places:  In the outermost Swift code for `description` and `debugDescription` (which return constant strings) and in the actual C formatting code.  (Because of the former, the latter is not technically necessary except to ensure that the C formatter is actually complete.)

tbkka commented on stdlib/public/core/Runtime.swift.gyb:
```diff
@@ -408,37 +408,17 @@ internal func _float${bits}ToString(
   _ value: Float${bits}, debug: Bool
```
I've kept the `debug` flag here even though it's not used.  The only difference between `debugDescription` and `description` is how they print NaNs.  That difference is now entirely encapsulated -- `description` checks for and returns the static string immediately; the bottom C layer always produces the debug form, so `debugDescription` can just pass NaNs down.

tbkka commented on stdlib/public/core/Runtime.swift.gyb:
```diff
@@ -408,37 +408,17 @@ internal func _float${bits}ToString(
+      let actualLength = _float${bits}ToStringImpl(bufferPtr, 32, value, debug)
+      return String._fromWellFormedCodeUnitSequence(
+                 UTF8.self,
+                 input: UnsafeBufferPointer(start: bufferPtr, count: Int(actualLength)))
```
Does anything need to change here to take advantage of @milseman's recent Small String work?

milseman replied:
Small string support hasn't been merged and I haven't come up with a clean alternative to `_fromWellFormedCodeUnitSequence` for them yet. When you merge, file a bug against me

tbkka replied:
Ideally, I'd like to have something like `String._unsafeFromWellFormedASCII(buffer:)` to use here.  If the buffer length is short, build and return a small ASCII string.  Otherwise, allocate and `memcpy` to build a long ASCII string.  In either case, omitting the charset shenanigans would be a sizable win.

milseman replied:
I have just what you need in the small-string branch: https://github.com/milseman/swift/blob/so_smol/stdlib/public/core/String.swift#L647

tbkka commented on stdlib/public/runtime/SwiftDtoa.c:
```diff
@@ -0,0 +1,2429 @@
+
+// Try to verify that the system floating-point types really are what we
+// expect.  Note that the code below is specific to these exact
+// floating-point representations.
```
These are mostly intended as simple guards in case someone copies this code elsewhere.  However, even within the Swift stdlib, we cannot safely assume that C `long double` really is Float80, so that check at least is important to have here.

stephentyrone replied:
Yup, that's fine.

tbkka commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -0,0 +1,828 @@
+#if !arch(i386)
+  // 32-bit i386 lacks signaling Double nans
+  expectNaN("snan", Double.signalingNaN)
+  expectNaN("-snan", -Double.signalingNaN)
```
@rintaro :  I'd like to add a test for `Double(nan: 0x4000000000001, signaling: true)`, but I'm not sure how that should print.  I think it should print as `"snan(0x4000000000001)"` but I think the previous code would print it as `"snan"` without the payload.

tbkka replied:
As far as I can tell, the standard does not reserve any bit for signaling NaN.  It only requires that the quiet bit is not set and the payload is not zero.

I would expect to reserve a single value as the default signaling payload (not a particular bit), so that `Double(nan: 0x4_0000_0000_0000, signaling: true).debugDescription` produces `"snan"` but `Double(nan: 0x4_0000_0000_0001, signaling: true).debugDescription` produces `"snan(0x4000000000001)"`.

stephentyrone replied:
> As far as I can tell, the standard does not reserve any bit for signaling NaN. It only requires that the quiet bit is not set and the payload is not zero.

It's weaker than that. There is no "quiet bit" in the standard. Some HW platforms use a "signaling bit" instead.

However, for platforms with the common-case behavior that the high-order significand bit is a quiet bit, Swift decided to canonicalize on `b10xxx...xxx` for qNaN and `b01xxx...xxx` for sNaN because this makes the same set of payloads encodable with each; otherwise quieting a sNaN is a lossy operation, which is undesirable.

(Note that we should actually switch to using `b11xxx...xxx` for qNaNs, so that `b00xxx...xxx` can serve as reserved bit patterns that are never generated by pure Swift operations on HW with the customary quiet bit).

tbkka committed:
Remove duplicate infinity handling

tbkka committed:
Use defined(__SIZEOF_INT128__) to detect uint128_t support

tbkka committed:
Separate `extracting` the integer part from `clearing` the integer part

The previous code was unnecessarily obfuscated by the attempt to combine
these two operations.

tbkka committed:
Use `UINT32_MAX` to mask off 32 bits of a larger integer

tbkka commented on stdlib/public/runtime/SwiftDtoa.c:
```diff
@@ -0,0 +1,2429 @@
+
+#include "swift/Runtime/SwiftDtoa.h"
+
+#if defined(__x86_64__) || defined(__aarch64__)
```
Done.

tbkka commented on stdlib/public/runtime/SwiftDtoa.c:
```diff
@@ -0,0 +1,2429 @@
+#define multiply128xi32(lhs, rhs) (*(lhs) *= (rhs))
+#define initialize128WithHigh64(dest, value) ((dest) = (__uint128_t)(value) << 64)
+#define extractHigh64From128(arg) ((uint64_t)((arg) >> 64))
+static int extractIntegerPart128(__uint128_t *fixed128, int fractionBits) {
```
I've split it into separate `extract` and `clear` operations to clarify.

tbkka commented on stdlib/public/runtime/SwiftDtoa.c:
```diff
@@ -0,0 +1,2429 @@
+    __uint128_t full = (__uint128_t)lhs * rhs;
+    return (uint64_t)(full >> 32);
+#else
+    static const uint64_t mask32 = ((uint64_t)1 << 32) - 1;
```
I've changed the spelling to `UINT32_MAX`.  I'll see about consolidating it as part of a follow-on.

tbkka commented on stdlib/public/runtime/SwiftDtoa.c:
```diff
@@ -0,0 +1,2429 @@
+    uint64_t t = (lhs & mask32) * (rhs & mask32);
+    t >>= 32;
+    uint64_t a = (lhs >> 32) * (rhs & mask32);
+    uint64_t b = (lhs & mask32) * (rhs >> 32);
```
Nice ideas!  I'll work through this in a follow-up PR.

stephentyrone commented:
What significant outstanding issues block this proceeding at this point?

tbkka commented:
@swift-ci Please test and merge

milseman commented:
@swift-ci please test

milseman commented:
Looks like you need to update `test/stdlib/PrintFloat.swift.gyb` for nan printing:

```
20:13:57 stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb, line 636
20:13:57 stdout>>> stacktrace:
20:13:57 stdout>>>   #0: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb:403
20:13:57 stdout>>> expected: "nan(0xffff)" (of type Swift.String)
20:13:57 stdout>>> actual: "nan" (of type Swift.String)
20:13:57 stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb, line 637
20:13:57 stdout>>> stacktrace:
20:13:57 stdout>>>   #0: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb:403
20:13:57 stdout>>> expected: "nan(0x3ffffffffffff)" (of type Swift.String)
20:13:57 stdout>>> actual: "nan" (of type Swift.String)
20:13:57 stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb, line 638
20:13:57 stdout>>> stacktrace:
20:13:57 stdout>>>   #0: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb:403
20:13:57 stdout>>> expected: "nan(0x3ffffffffffff)" (of type Swift.String)
20:13:57 stdout>>> actual: "nan" (of type Swift.String)
20:13:57 stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb, line 640
20:13:57 stdout>>> stacktrace:
20:13:57 stdout>>>   #0: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/PrintFloat.swift.gyb:403
20:13:57 stdout>>> expected: "-nan(0xffff)" (of type Swift.String)
20:13:57 stdout>>> actual: "-nan" (of type Swift.String)
```

tbkka committed:
Correct the expected NaN results for 32-bit i386

tbkka commented:
@milseman: Thanks!  It took a while to get the tests running, but I've verified the corrected results against the 32-bit i386 iPhone simulator.  I'd like to verify it against a real 32-bit ARM device as well; any hints?

tbkka committed:
Merge branch 'master' into tbkka-floating-point-printing-C

tbkka commented:
@swift-ci Please smoke test

stephentyrone commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -631,13 +631,23 @@ PrintTests.test("Printable_Double") {
+  expectNaN("nan", Double(nan: 65535, signaling: false))
+  expectNaN("nan", Float64(bitPattern: 0x7fff_ffff_ffff_ffff))
+  expectNaN("nan", Float64(bitPattern: 0x7ffb_ffff_ffff_ffff))
+  expectNaN("-nan", -Double(nan: 65535, signaling: false))
```
I expect this test is going to be fragile w.r.t. optimization on arm, where we run with the "Default NaN" bit set in FPCR, as that bit means that if the negation actually happens at runtime the resulting NaN won't be "negative".

More generally, IEEE 754 says that we shouldn't "interpret" the signbit of NaN (except for four operations: copy, copysign, negate, abs), so probably the best fix here is to simply print all NaNs as "nan" rather than "[-]nan". AFAIK we were only ever printing the leading `-` because it happened to fall out of `snprintf`.

xwu replied:
These are testing the _debug_ description, where sign, quiet bit, and payload have all been printed in the past—IMO that’s useful for debugging particularly because of platform-specific differences in NaN handling.

stephentyrone replied:
I don't think it's actually very useful--its information about the encoding, not the value, which is  best accessed via `x.bitPattern`. My main concern though is that it's challenging to write portable tests for these cases.

I *do* think that we should be printing `snan` for signaling NaNs, though. And we should print the payload in the debug description.

xwu replied:
It’s of course possible to access everything about a NaN through its bit pattern, but FWIW I do use debug description to debug the sign of NaN. My point is that I find the result meaningful for debugging precisely _because_ it’s challenging to write portable tests.

I think if we’re going to debug print signaling vs. quiet and payload, then we’re wise to debug print the sign; after all, these are all aspects of NaN considered in the IEEE totalOrder operation.

stephentyrone replied:
Printing `snan` is important because you're supposed to either signal invalid or print a different string--otherwise the signaling information is lost--and Swift doesn't model fenv, so we can't signal invalid. Printing the payload is less vital, but possibly useful for some people.

Printing the sign is of some interest to us as stdlib developers, but one else *should* care about it. It only has semantic meaning in `copysign` and the corresponding `init` and the totalOrder operation, which almost no one uses =)

I'm not going to say that we absolutely shouldn't print it, just that I don't think it carries its weight, and that it's leading us toward writing overly-restrictive tests.

stephentyrone commented:
I think that this warrants some explanation in CHANGELOG.md.

tbkka commented:
> I think that this warrants some explanation in CHANGELOG.md.

Do you mean the debug printing of NaNs?  It's not a change; I duplicated the behavior of the existing code.  (I did add tests but as far as I know, those are just verifying existing behavior.)

As for "portable tests", I agree that the NaN handling does complicate things.  I'm not worried _yet_ about test complication -- there are already plenty of conditionals for Float80 support and a few more aren't a big headache.  If it becomes burdensome, we can either relax the tests or simplify the format at that time.

stephentyrone commented:
No, I mean the printing changes in general; I don’t think you need to mention NaNs.

tbkka commented:
@swift-ci Please smoke test

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
Linux failure looks weird:
```
/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/atomic:185:7: error: exception specification of explicitly defaulted default constructor does not match the calculated one
12:35:39       atomic() noexcept = default;
12:35:39       ^
12:35:39 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/stdlib/public/runtime/WeakReference.h:151:36: note: in instantiation of template class 'std::atomic<swift::WeakReferenceBits>' requested here
12:35:39     std::atomic<WeakReferenceBits> nativeValue;
12:35:39                                    ^
12:35:39 1 error generated.
```

@shahmishal is this fallout from a version update? It's a possible difference between GCC and Clang.

milseman commented:
@gparker42 any idea here?

milseman commented:
It's weird because this code hasn't changed in over a year.

milseman commented:
@swift-ci please smoke benchmark

(since small strings got in and could influence some workloads)

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please at least tell me if you're going to ignore me

milseman commented:
@shahmishal it seems like @swift-ci hasn't been smoke benchmarking since 6am yesterday. Is it available for response?

shahmishal commented:
@swift-ci smoke benchmark

milseman commented:
Ah, I see, I should be less polite when @swift-ci comes back from its smoke break.

shahmishal commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke test Linux Platform

tbkka commented:
Wow.  Looks like both changes (Small strings and the new formatter) had big effects, especially for single precision floats:  Float_description_small (which tests numbers around 1.0) went from 41747 with neither to 4626 with both optimizations, assuming I'm reading all of this right.  (4626 usec per 100,000 conversions is around 46ns total time to convert a Float into a String.)

But even for Float80_description_uniform, the small-string code seems to have made an additional 30% improvement.  That surprises me, since these values should not be short enough to fit into a small string.

tbkka committed:
Merge branch 'master' into tbkka-floating-point-printing-C

tbkka committed:
Merge branch 'tbkka-floating-point-printing-C' of github.com:tbkka/swift into tbkka-floating-point-printing-C

tbkka commented:
Seems the Linux failure here might be a clang bug?
       https://lists.llvm.org/pipermail/llvm-bugs/2017-June/056371.html

tbkka commented:
I merged master again, maybe it will have some impact...

@swift-ci Please smoke test

stephentyrone commented:
@swift-ci Please smoke test linux

milseman commented:
@swift-ci please clean smoke test linux platform

stephentyrone commented:
Please smoke test harder.

milseman commented:
Other builds on the exact same machine don't hit this error...

milseman commented:
@jckarter any idea why there would be the following failure on Linux?

```
09:52:46 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/llvm-linux-x86_64/./bin/clang++   -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Istdlib/public/runtime -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/stdlib/public/runtime -Iinclude -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/include -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/llvm/include -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/llvm-linux-x86_64/include -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/llvm-linux-x86_64/tools/clang/include -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/llvm/tools/clang/include -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/cmark/src -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/cmark-linux-x86_64/src -isystem /usr/include/x86_64-linux-gnu -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wcovered-switch-default -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fcolor-diagnostics -ffunction-sections -fdata-sections -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -fno-sanitize=all -DLLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING=1 -DSWIFT_RUNTIME_ENABLE_GUARANTEED_NORMAL_ARGUMENTS=1 -O3    -UNDEBUG  -Wall -Wglobal-constructors -Wexit-time-destructors -fvisibility=hidden -D__SWIFT_CURRENT_DYLIB=swiftCore -DswiftCore_EXPORTS -I/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/include -target x86_64-unknown-linux-gnu -O2 -g0 -DNDEBUG -MMD -MT stdlib/public/runtime/CMakeFiles/swiftRuntime-linux-x86_64.dir/SwiftObject.mm.o -MF stdlib/public/runtime/CMakeFiles/swiftRuntime-linux-x86_64.dir/SwiftObject.mm.o.d -o stdlib/public/runtime/CMakeFiles/swiftRuntime-linux-x86_64.dir/SwiftObject.mm.o -c /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/stdlib/public/runtime/SwiftObject.mm
09:52:46 In file included from /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/stdlib/public/runtime/SwiftObject.mm:29:
09:52:46 In file included from /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/include/swift/Runtime/Casting.h:20:
09:52:46 In file included from /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/include/swift/Runtime/Metadata.h:20:
09:52:46 
/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/atomic:185:7: error: exception specification of explicitly defaulted default constructor does not match the calculated one
09:52:46       atomic() noexcept = default;
09:52:46       ^
09:52:46 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/stdlib/public/runtime/WeakReference.h:151:36: note: in instantiation of template class 'std::atomic<swift::WeakReferenceBits>' requested here
09:52:46     std::atomic<WeakReferenceBits> nativeValue;
09:52:46                                    ^
09:52:46 1 error generated.
```

jckarter commented:
@milseman You may need to put `noexcept` on `WeakReferenceBits`' own default constructor. Maybe `libstdc++` marks `atomic()` as noexcept but `libc++` doesn't?

tbkka committed:
Make the C++ exceptions here consistent

Adding a C source file somehow exposed an issue in an unrelated C++ file.
Thanks to Joe Groff for the fix.

tbkka commented:
Interesting:  Adding any C file to stdlib/public/runtime/CMakeLists.txt will unmask this.

Some quick tests suggest that @jckarter's suggested change fixes it.

tbkka commented:
@swift-ci Please smoke test

tbkka commented:
Now the Linux build fails at a different point:
```
17:44:03 FAILED: tests/dispatch_starfish 
17:44:03 : && /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/llvm-linux-x86_64/bin/clang  -O3 -DNDEBUG  -lrt tests/CMakeFiles/dispatch_starfish.dir/dispatch_starfish.c.o  -o tests/dispatch_starfish  src/libdispatch.so libBlocksRuntime.a -lbsd /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64/libswiftCore.so /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64/libswiftSwiftOnoneSupport.so tests/libbsdtests.a -pthread -Wl,-rpath,/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/libdispatch-linux-x86_64/src:/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64 && :
17:44:03 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/lib/swift/linux/x86_64/libswiftCore.so: undefined reference to `__gnu_objc_personality_v0'
17:44:03 
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

tbkka committed:
Rename SwiftDtoa to ".cpp"

Having a C file in stdlib/public/runtime causes strange
build failures on Linux in unrelated C++ files.

As a workaround, rename SwiftDtoa.c to .cpp to see
if that avoids the problems.

tbkka commented:
I've renamed SwiftDtoa.c to SwiftDtoa.cpp to see if that avoids the Linux problems.

tbkka commented:
@swift-ci Please smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci smoke test

stephentyrone commented:
@milseman @tbkka I would like to merge this and pursue the minor issues I raised in separate PRs. Any objections?

tbkka committed:
Revert "Make the C++ exceptions here consistent"

This reverts commit 6cd5c2056690572a41366c2f1837059e64ca17c6.

tbkka commented:
I've backed out 6cd5c20, since it's unrelated.   (I think it's probably a good change, but should be handled separately.)

With that, as soon as the smoke tests look good, merge away!

tbkka commented:
@swift-ci smoke test

milseman commented on stdlib/public/runtime/SwiftDtoa.cpp:
```diff
@@ -9,6 +9,11 @@
+//
+// Note: This is really a C file, but Swift's build system for Linux is
+// partially allergic to C, so it's being compiled as ".cpp" for now.  Please
+// don't infect it with C++-isms.
```
Why?

tbkka replied:
"Don't infect with C++-isms":  Because I expect to have some discussions soon about using this code in other projects that may not be able to easily use C++.  Keeping this pure C (at least for now) will simplify possible reuse.

Plus, my test suites are predominantly pure C right now and I have a number of simplifications and/or performance improvements to the code that I'm actively testing.

milseman reviewed (commented):
Just curious

milseman commented:
@swift-ci please smoke test
